{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "287\n"
     ]
    }
   ],
   "source": [
    "path = 'gwas-association-downloaded_2018-08-22-jz liu.tsv'\n",
    "liu = pd.read_csv(path, sep='\\t')\n",
    "path = 'gwas-association-downloaded_2018-08-22-de lange.tsv'\n",
    "kdl = pd.read_csv(path, sep='\\t')\n",
    "\n",
    "both = pd.concat([liu, kdl]).reset_index(drop=True)\n",
    "ibd = both[both['DISEASE/TRAIT'] == 'Inflammatory bowel disease']\n",
    "ibd = ibd.drop_duplicates(subset=['CHR_ID', 'CHR_POS']).reset_index(drop=True)\n",
    "print(len(ibd))\n",
    "# threshold_significance = 5e-6\n",
    "# print(sum(ibd['P-VALUE'] < threshold_significance))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = 'QuickGO-annotations-1535039825731-20180823.tsv'\n",
    "annotations = pd.read_csv(path, sep='\\t')\n",
    "annotations = annotations[annotations['QUALIFIER'] == 'part_of']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_dashes(x):\n",
    "    if pd.isnull(x):\n",
    "        return x\n",
    "    if ' - ' in x:\n",
    "        return x.split(' - ')\n",
    "    return [x]\n",
    "\n",
    "def remove_commas(x):\n",
    "    r = []\n",
    "    if not isinstance(x, list) and pd.isnull(x):\n",
    "        return x\n",
    "    for i, y in enumerate(x):\n",
    "\n",
    "        if ', ' in y:\n",
    "             r += y.split(', ')\n",
    "        else:\n",
    "            r.append(y)\n",
    "    return r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MAPPED_GENE_extracted</th>\n",
       "      <th>MAPPED_GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[IL23R, RNU4ATAC4P]</td>\n",
       "      <td>IL23R - RNU4ATAC4P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[LOC105374736, LOC105374737]</td>\n",
       "      <td>LOC105374736 - LOC105374737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[LOC101927300, LINC01475]</td>\n",
       "      <td>LOC101927300 - LINC01475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[HLA-DQA1, HLA-DQB1]</td>\n",
       "      <td>HLA-DQA1 - HLA-DQB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[CARD9]</td>\n",
       "      <td>CARD9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          MAPPED_GENE_extracted                  MAPPED_GENE\n",
       "0           [IL23R, RNU4ATAC4P]           IL23R - RNU4ATAC4P\n",
       "1  [LOC105374736, LOC105374737]  LOC105374736 - LOC105374737\n",
       "2     [LOC101927300, LINC01475]     LOC101927300 - LINC01475\n",
       "3          [HLA-DQA1, HLA-DQB1]          HLA-DQA1 - HLA-DQB1\n",
       "4                       [CARD9]                        CARD9"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ibd['MAPPED_GENE_extracted'] = ibd['MAPPED_GENE'].apply(remove_dashes)\n",
    "ibd['MAPPED_GENE_extracted'] = ibd['MAPPED_GENE_extracted'].apply(remove_commas)\n",
    "ibd[['MAPPED_GENE_extracted', 'MAPPED_GENE']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "MAPPED_GENE_extracted = set()\n",
    "for x in ibd['MAPPED_GENE_extracted']:\n",
    "    if not isinstance(x, list) and pd.isnull(x):\n",
    "        continue\n",
    "    MAPPED_GENE_extracted.update(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACO2',\n",
       " 'ADAD1',\n",
       " 'ADAM30',\n",
       " 'ADCY3',\n",
       " 'ADCY7',\n",
       " 'AIMP1P2',\n",
       " 'AKAP11',\n",
       " 'ALDH7A1P4',\n",
       " 'AMZ1',\n",
       " 'ANKRD55',\n",
       " 'ARL2BPP4',\n",
       " 'ARPC2',\n",
       " 'ATG16L1',\n",
       " 'ATP6V1G3',\n",
       " 'ATXN2',\n",
       " 'BACH2',\n",
       " 'BANK1',\n",
       " 'BOK-AS1',\n",
       " 'BSN',\n",
       " 'BTBD8',\n",
       " 'BTF3L4P3',\n",
       " 'C11orf30',\n",
       " 'C17orf67',\n",
       " 'C1orf106',\n",
       " 'C5orf56',\n",
       " 'C5orf66',\n",
       " 'C7orf72',\n",
       " 'CAMK2A',\n",
       " 'CAMK2G',\n",
       " 'CARD9',\n",
       " 'CCDC26',\n",
       " 'CCL2',\n",
       " 'CCL20',\n",
       " 'CCL7',\n",
       " 'CCND3',\n",
       " 'CCND3P1',\n",
       " 'CD226',\n",
       " 'CD28',\n",
       " 'CD40',\n",
       " 'CD6',\n",
       " 'CDC37',\n",
       " 'CDC42SE2',\n",
       " 'CDKAL1',\n",
       " 'CEBPA',\n",
       " 'CELSR3',\n",
       " 'CIT',\n",
       " 'COMMD7',\n",
       " 'COX6B1P6',\n",
       " 'CPEB4',\n",
       " 'CPSF3L',\n",
       " 'CREM',\n",
       " 'CRTC3',\n",
       " 'CRTC3-AS1',\n",
       " 'CTIF',\n",
       " 'CUL1',\n",
       " 'CUL2',\n",
       " 'CYCSP42',\n",
       " 'DAP',\n",
       " 'DAP3',\n",
       " 'DENND1B',\n",
       " 'DNMT3B',\n",
       " 'DPH5',\n",
       " 'DUSP16',\n",
       " 'DUSP22',\n",
       " 'EEF1A1P27',\n",
       " 'EPHB4',\n",
       " 'ERAP1',\n",
       " 'ERAP2',\n",
       " 'ERGIC1',\n",
       " 'ETS1',\n",
       " 'EXOC6',\n",
       " 'F5',\n",
       " 'FABP3P2',\n",
       " 'FADS2',\n",
       " 'FAM171B',\n",
       " 'FAM53B',\n",
       " 'FCGR2A',\n",
       " 'FERMT1',\n",
       " 'FGFR1OP2P1',\n",
       " 'FLJ31356',\n",
       " 'FOS',\n",
       " 'FOSL2',\n",
       " 'FOXP1',\n",
       " 'FUT2',\n",
       " 'GALC',\n",
       " 'GCKR',\n",
       " 'GLYAT',\n",
       " 'GNA12',\n",
       " 'GPBAR1',\n",
       " 'GPR18',\n",
       " 'GPR183',\n",
       " 'GPR35',\n",
       " 'GPR65',\n",
       " 'GRB7',\n",
       " 'GRID2IP',\n",
       " 'HDAC7',\n",
       " 'HGFAC',\n",
       " 'HHEX',\n",
       " 'HIF3A',\n",
       " 'HLA-DQA1',\n",
       " 'HLA-DQB1',\n",
       " 'HNF4A',\n",
       " 'HNRNPA1P41',\n",
       " 'HORMAD2',\n",
       " 'IFIH1',\n",
       " 'IFNGR2',\n",
       " 'IKZF1',\n",
       " 'IKZF3',\n",
       " 'IL10',\n",
       " 'IL18R1',\n",
       " 'IL18RAP',\n",
       " 'IL1RL1',\n",
       " 'IL2',\n",
       " 'IL23R',\n",
       " 'IL27',\n",
       " 'IL2RA',\n",
       " 'IMPG2',\n",
       " 'IPMK',\n",
       " 'IRF4',\n",
       " 'IRF5',\n",
       " 'IRF6',\n",
       " 'IRGM',\n",
       " 'ITGA4',\n",
       " 'ITGAL',\n",
       " 'ITGB8',\n",
       " 'ITLN1',\n",
       " 'JAK2',\n",
       " 'KIAA1109',\n",
       " 'KIF3B',\n",
       " 'KIR3DL2',\n",
       " 'KPNA7',\n",
       " 'KRT18P56',\n",
       " 'KSR1',\n",
       " 'LACC1',\n",
       " 'LINC00332',\n",
       " 'LINC00484',\n",
       " 'LINC00581',\n",
       " 'LINC00598',\n",
       " 'LINC00824',\n",
       " 'LINC01271',\n",
       " 'LINC01430',\n",
       " 'LINC01475',\n",
       " 'LITAF',\n",
       " 'LOC100130299',\n",
       " 'LOC100287157',\n",
       " 'LOC100420500',\n",
       " 'LOC100421782',\n",
       " 'LOC100533842',\n",
       " 'LOC100996324',\n",
       " 'LOC100996583',\n",
       " 'LOC101060363',\n",
       " 'LOC101927156',\n",
       " 'LOC101927239',\n",
       " 'LOC101927300',\n",
       " 'LOC101927745',\n",
       " 'LOC101927897',\n",
       " 'LOC101928354',\n",
       " 'LOC101928813',\n",
       " 'LOC101930164',\n",
       " 'LOC102659288',\n",
       " 'LOC102723415',\n",
       " 'LOC102723649',\n",
       " 'LOC105369736',\n",
       " 'LOC105370177',\n",
       " 'LOC105370571',\n",
       " 'LOC105371082',\n",
       " 'LOC105371388',\n",
       " 'LOC105371389',\n",
       " 'LOC105371473',\n",
       " 'LOC105371618',\n",
       " 'LOC105371887',\n",
       " 'LOC105372657',\n",
       " 'LOC105372877',\n",
       " 'LOC105373831',\n",
       " 'LOC105374409',\n",
       " 'LOC105374410',\n",
       " 'LOC105374411',\n",
       " 'LOC105374541',\n",
       " 'LOC105374736',\n",
       " 'LOC105374737',\n",
       " 'LOC105374780',\n",
       " 'LOC105375015',\n",
       " 'LOC105375444',\n",
       " 'LOC105375466',\n",
       " 'LOC105375746',\n",
       " 'LOC105376491',\n",
       " 'LOC105376823',\n",
       " 'LOC105376856',\n",
       " 'LOC105376976',\n",
       " 'LOC105377068',\n",
       " 'LOC105377139',\n",
       " 'LOC105378049',\n",
       " 'LOC105378083',\n",
       " 'LOC105378120',\n",
       " 'LOC105378482',\n",
       " 'LOC105379031',\n",
       " 'LOC105447645',\n",
       " 'LOC107983952',\n",
       " 'LOC107984299',\n",
       " 'LOC107984366',\n",
       " 'LOC107984526',\n",
       " 'LOC107984569',\n",
       " 'LOC107984725',\n",
       " 'LOC107984739',\n",
       " 'LOC107984914',\n",
       " 'LOC107985484',\n",
       " 'LOC107985575',\n",
       " 'LOC107986449',\n",
       " 'LOC107986522',\n",
       " 'LOC107986572',\n",
       " 'LOC285626',\n",
       " 'LOC285627',\n",
       " 'LOC285740',\n",
       " 'LOC340090',\n",
       " 'LOC392145',\n",
       " 'LOC392787',\n",
       " 'LOC442263',\n",
       " 'LOC553103',\n",
       " 'LOC645266',\n",
       " 'LOC645481',\n",
       " 'LOC646347',\n",
       " 'LPP',\n",
       " 'LRRK2',\n",
       " 'LSP1',\n",
       " 'LTBR',\n",
       " 'MAGOH3P',\n",
       " 'MAML2',\n",
       " 'MAP3K8',\n",
       " 'METTL10',\n",
       " 'MIR3939',\n",
       " 'MRPS35P3',\n",
       " 'MST1',\n",
       " 'NDFIP1',\n",
       " 'NFATC1',\n",
       " 'NIFKP1',\n",
       " 'NOD2',\n",
       " 'NOTCH1',\n",
       " 'NR5A2',\n",
       " 'NXPE1',\n",
       " 'NXPE2',\n",
       " 'OSMR',\n",
       " 'PARK7',\n",
       " 'PDE4A',\n",
       " 'PDGFB',\n",
       " 'PHACTR2',\n",
       " 'PIGCP2',\n",
       " 'PLA2G4A',\n",
       " 'PLA2R1',\n",
       " 'PLAU',\n",
       " 'PLCG2',\n",
       " 'PNKD',\n",
       " 'POFUT1',\n",
       " 'PPP5C',\n",
       " 'PRDM1',\n",
       " 'PRKCB',\n",
       " 'PTGIR',\n",
       " 'PTK2B',\n",
       " 'PTPN2',\n",
       " 'PTPRC',\n",
       " 'PTPRK',\n",
       " 'PTRF',\n",
       " 'PUS10',\n",
       " 'RELA',\n",
       " 'RGS14',\n",
       " 'RN7SKP211',\n",
       " 'RNASET2',\n",
       " 'RNF186',\n",
       " 'RNU1-150P',\n",
       " 'RNU4ATAC4P',\n",
       " 'RNU6-222P',\n",
       " 'RNU6-925P',\n",
       " 'RNY3P9',\n",
       " 'RORC',\n",
       " 'RPL10AP8',\n",
       " 'RPL13P2',\n",
       " 'RPL23AP12',\n",
       " 'RPL3',\n",
       " 'RPL32P17',\n",
       " 'RPS21P8',\n",
       " 'RPS6KA4',\n",
       " 'RTEL1-TNFRSF6B',\n",
       " 'SBNO2',\n",
       " 'SCAMP3',\n",
       " 'SEPHS2',\n",
       " 'SH2B1',\n",
       " 'SH2B3',\n",
       " 'SKAP2',\n",
       " 'SLAIN2',\n",
       " 'SLC22A23',\n",
       " 'SLC22A4',\n",
       " 'SLC2A13',\n",
       " 'SLC39A11',\n",
       " 'SLC7A10',\n",
       " 'SMAD3',\n",
       " 'SMAD7',\n",
       " 'SMNDC1',\n",
       " 'SNRPGP8',\n",
       " 'SP140',\n",
       " 'SPRED2',\n",
       " 'STAT4',\n",
       " 'STAT5A',\n",
       " 'SUFU',\n",
       " 'SYNGR1',\n",
       " 'TAB2',\n",
       " 'TET2',\n",
       " 'TEX41',\n",
       " 'THADA',\n",
       " 'THEMIS',\n",
       " 'TM9SF4',\n",
       " 'TMBIM1',\n",
       " 'TNFRSF6B',\n",
       " 'TNFSF15',\n",
       " 'TNFSF8',\n",
       " 'TOM1',\n",
       " 'TPT1P10',\n",
       " 'TRAF3IP2',\n",
       " 'TRAF3IP2-AS1',\n",
       " 'TRIB1',\n",
       " 'TSPAN14',\n",
       " 'TUBD1',\n",
       " 'TUFM',\n",
       " 'TYK2',\n",
       " 'UBAC2',\n",
       " 'UBE2L3',\n",
       " 'YBX1P5',\n",
       " 'ZFP36L1',\n",
       " 'ZFP91-CNTF',\n",
       " 'ZMIZ1',\n",
       " 'ZNF300',\n",
       " 'ZNF300P1',\n",
       " 'ZNF365',\n",
       " 'ZNF831',\n",
       " 'ZPBP2'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MAPPED_GENE_extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       AIM2\n",
       "1       AIM2\n",
       "2       AIM2\n",
       "3      CASP1\n",
       "4      CASP1\n",
       "5      CASP1\n",
       "6      CASP1\n",
       "7      CASP1\n",
       "8      CASP1\n",
       "9      CASP4\n",
       "10     CASP4\n",
       "11     CASP4\n",
       "12     CASP4\n",
       "13     CASP4\n",
       "17     CASP5\n",
       "18     GSDMD\n",
       "19     GSDMD\n",
       "20     GSDMD\n",
       "21     GSDMD\n",
       "22     NLRP6\n",
       "23     NLRP6\n",
       "27     NLRP9\n",
       "28     NLRP9\n",
       "29     NLRP9\n",
       "30     NLRP3\n",
       "31     NLRP3\n",
       "32     NLRP3\n",
       "33     NLRP3\n",
       "34     NLRP1\n",
       "35     NLRP1\n",
       "36     NLRP1\n",
       "37     NLRP1\n",
       "38     DHX33\n",
       "39     DHX33\n",
       "40     DHX33\n",
       "41     NLRC4\n",
       "42     NLRC4\n",
       "43     NLRC4\n",
       "44    PYCARD\n",
       "45    PYCARD\n",
       "46    PYCARD\n",
       "47     CARD8\n",
       "Name: SYMBOL, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations.SYMBOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(annotations.SYMBOL) & MAPPED_GENE_extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>P-VALUE (TEXT)</th>\n",
       "      <th>OR or BETA</th>\n",
       "      <th>95% CI (TEXT)</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>CNV</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "      <th>MAPPED_GENE_extracted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2016-07-24</td>\n",
       "      <td>26192919</td>\n",
       "      <td>Liu JZ</td>\n",
       "      <td>2015-07-20</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/26192919</td>\n",
       "      <td>Association analyses identify 38 susceptibilit...</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>12,882 European ancestry cases, 21,770 Europea...</td>\n",
       "      <td>25,273 European ancestry cases, 26,715 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Affymetrix, Illumina [~ 9000000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>inflammatory bowel disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003767</td>\n",
       "      <td>GCST003043</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>[IL23R, RNU4ATAC4P]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2016-07-24</td>\n",
       "      <td>26192919</td>\n",
       "      <td>Liu JZ</td>\n",
       "      <td>2015-07-20</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/26192919</td>\n",
       "      <td>Association analyses identify 38 susceptibilit...</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>12,882 European ancestry cases, 21,770 Europea...</td>\n",
       "      <td>25,273 European ancestry cases, 26,715 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Affymetrix, Illumina [~ 9000000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>inflammatory bowel disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003767</td>\n",
       "      <td>GCST003043</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>[LOC105374736, LOC105374737]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2016-07-24</td>\n",
       "      <td>26192919</td>\n",
       "      <td>Liu JZ</td>\n",
       "      <td>2015-07-20</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/26192919</td>\n",
       "      <td>Association analyses identify 38 susceptibilit...</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>12,882 European ancestry cases, 21,770 Europea...</td>\n",
       "      <td>25,273 European ancestry cases, 26,715 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Affymetrix, Illumina [~ 9000000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>inflammatory bowel disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003767</td>\n",
       "      <td>GCST003043</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>[LOC101927300, LINC01475]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2016-07-24</td>\n",
       "      <td>26192919</td>\n",
       "      <td>Liu JZ</td>\n",
       "      <td>2015-07-20</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/26192919</td>\n",
       "      <td>Association analyses identify 38 susceptibilit...</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>12,882 European ancestry cases, 21,770 Europea...</td>\n",
       "      <td>25,273 European ancestry cases, 26,715 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>1.213039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Affymetrix, Illumina [~ 9000000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>inflammatory bowel disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003767</td>\n",
       "      <td>GCST003043</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>[HLA-DQA1, HLA-DQB1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2016-07-24</td>\n",
       "      <td>26192919</td>\n",
       "      <td>Liu JZ</td>\n",
       "      <td>2015-07-20</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/26192919</td>\n",
       "      <td>Association analyses identify 38 susceptibilit...</td>\n",
       "      <td>Inflammatory bowel disease</td>\n",
       "      <td>12,882 European ancestry cases, 21,770 Europea...</td>\n",
       "      <td>25,273 European ancestry cases, 26,715 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Affymetrix, Illumina [~ 9000000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>inflammatory bowel disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003767</td>\n",
       "      <td>GCST003043</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>[CARD9]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATE ADDED TO CATALOG  PUBMEDID FIRST AUTHOR        DATE    JOURNAL  \\\n",
       "0            2016-07-24  26192919       Liu JZ  2015-07-20  Nat Genet   \n",
       "1            2016-07-24  26192919       Liu JZ  2015-07-20  Nat Genet   \n",
       "2            2016-07-24  26192919       Liu JZ  2015-07-20  Nat Genet   \n",
       "3            2016-07-24  26192919       Liu JZ  2015-07-20  Nat Genet   \n",
       "4            2016-07-24  26192919       Liu JZ  2015-07-20  Nat Genet   \n",
       "\n",
       "                                   LINK  \\\n",
       "0  www.ncbi.nlm.nih.gov/pubmed/26192919   \n",
       "1  www.ncbi.nlm.nih.gov/pubmed/26192919   \n",
       "2  www.ncbi.nlm.nih.gov/pubmed/26192919   \n",
       "3  www.ncbi.nlm.nih.gov/pubmed/26192919   \n",
       "4  www.ncbi.nlm.nih.gov/pubmed/26192919   \n",
       "\n",
       "                                               STUDY  \\\n",
       "0  Association analyses identify 38 susceptibilit...   \n",
       "1  Association analyses identify 38 susceptibilit...   \n",
       "2  Association analyses identify 38 susceptibilit...   \n",
       "3  Association analyses identify 38 susceptibilit...   \n",
       "4  Association analyses identify 38 susceptibilit...   \n",
       "\n",
       "                DISEASE/TRAIT  \\\n",
       "0  Inflammatory bowel disease   \n",
       "1  Inflammatory bowel disease   \n",
       "2  Inflammatory bowel disease   \n",
       "3  Inflammatory bowel disease   \n",
       "4  Inflammatory bowel disease   \n",
       "\n",
       "                                 INITIAL SAMPLE SIZE  \\\n",
       "0  12,882 European ancestry cases, 21,770 Europea...   \n",
       "1  12,882 European ancestry cases, 21,770 Europea...   \n",
       "2  12,882 European ancestry cases, 21,770 Europea...   \n",
       "3  12,882 European ancestry cases, 21,770 Europea...   \n",
       "4  12,882 European ancestry cases, 21,770 Europea...   \n",
       "\n",
       "                             REPLICATION SAMPLE SIZE  \\\n",
       "0  25,273 European ancestry cases, 26,715 Europea...   \n",
       "1  25,273 European ancestry cases, 26,715 Europea...   \n",
       "2  25,273 European ancestry cases, 26,715 Europea...   \n",
       "3  25,273 European ancestry cases, 26,715 Europea...   \n",
       "4  25,273 European ancestry cases, 26,715 Europea...   \n",
       "\n",
       "               ...              P-VALUE (TEXT)  OR or BETA  95% CI (TEXT)  \\\n",
       "0              ...                        (EA)         NaN            NaN   \n",
       "1              ...                        (EA)         NaN            NaN   \n",
       "2              ...                        (EA)         NaN            NaN   \n",
       "3              ...                        (EA)    1.213039            NaN   \n",
       "4              ...                        (EA)         NaN            NaN   \n",
       "\n",
       "                   PLATFORM [SNPS PASSING QC] CNV                MAPPED_TRAIT  \\\n",
       "0  Affymetrix, Illumina [~ 9000000] (imputed)   N  inflammatory bowel disease   \n",
       "1  Affymetrix, Illumina [~ 9000000] (imputed)   N  inflammatory bowel disease   \n",
       "2  Affymetrix, Illumina [~ 9000000] (imputed)   N  inflammatory bowel disease   \n",
       "3  Affymetrix, Illumina [~ 9000000] (imputed)   N  inflammatory bowel disease   \n",
       "4  Affymetrix, Illumina [~ 9000000] (imputed)   N  inflammatory bowel disease   \n",
       "\n",
       "                       MAPPED_TRAIT_URI STUDY ACCESSION  \\\n",
       "0  http://www.ebi.ac.uk/efo/EFO_0003767      GCST003043   \n",
       "1  http://www.ebi.ac.uk/efo/EFO_0003767      GCST003043   \n",
       "2  http://www.ebi.ac.uk/efo/EFO_0003767      GCST003043   \n",
       "3  http://www.ebi.ac.uk/efo/EFO_0003767      GCST003043   \n",
       "4  http://www.ebi.ac.uk/efo/EFO_0003767      GCST003043   \n",
       "\n",
       "          GENOTYPING TECHNOLOGY         MAPPED_GENE_extracted  \n",
       "0  Genome-wide genotyping array           [IL23R, RNU4ATAC4P]  \n",
       "1  Genome-wide genotyping array  [LOC105374736, LOC105374737]  \n",
       "2  Genome-wide genotyping array     [LOC101927300, LINC01475]  \n",
       "3  Genome-wide genotyping array          [HLA-DQA1, HLA-DQB1]  \n",
       "4  Genome-wide genotyping array                       [CARD9]  \n",
       "\n",
       "[5 rows x 39 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ibd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import re\n",
    "url_server = 'https://rest.ensembl.org'\n",
    "re_location = re.compile(r'(?P<chrom>\\w+):(?P<start>\\d+)-(?P<end>\\d+)')\n",
    "\n",
    "\n",
    "def extract_chrom_start_end(location):\n",
    "    m = re_location.match(location)\n",
    "    return m.groupdict()\n",
    "\n",
    "\n",
    "def url_get(url):\n",
    "#     print(url)\n",
    "    r = requests.get(url, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        raise Exception()\n",
    "    return r.json()\n",
    "\n",
    "\n",
    "def parse_children(child):\n",
    "    results = []\n",
    "    if 'children' in child:\n",
    "        for children in child['children']:\n",
    "            results += parse_children(children)\n",
    "    if 'sequence' in child:\n",
    "        c = child['sequence']\n",
    "        if 'name' in c:\n",
    "            name = c['name']\n",
    "            location = c['location']\n",
    "            info = extract_chrom_start_end(location)\n",
    "            assert len(c['id']) ==  1\n",
    "            info['id'] = c['id'][0]['accession']\n",
    "            info['name'] = c['name']\n",
    "            results.append(info)\n",
    "    return results\n",
    "\n",
    "\n",
    "def get_region_info_for_gene(gene):\n",
    "    \"\"\"\n",
    "    there is a chance this will return > 1 regions for 1 gene - process both in the pipeline\n",
    "    \"\"\"\n",
    "    url_endpoint = '/genetree/member/symbol/homo_sapiens/'\n",
    "    url_options = '?prune_species=homo_sapiens;content-type=application/json'\n",
    "    url = ''.join([url_server, url_endpoint, gene, url_options])\n",
    "    data = url_get(url)\n",
    "    \n",
    "    \"\"\"\n",
    "    try:\n",
    "    \n",
    "    except e:\n",
    "        filter on name to include gene name ie \"IL10RB-001\"\n",
    "    \"\"\"\n",
    "#     return data\n",
    "    tree = data['tree']\n",
    "    if 'sequence' in tree:\n",
    "        location = tree['sequence']['location']\n",
    "        info = extract_chrom_start_end(location)\n",
    "        info['id'] = tree['id']['accession']\n",
    "    else:\n",
    "        info = parse_children(tree)\n",
    "        info = [x for x in info if f'{gene}-' in x['name']]\n",
    "        assert len(info) == 1\n",
    "    if isinstance(info, list):\n",
    "        return info[0]\n",
    "    else:\n",
    "        return info\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "testing = ibd.copy()\n",
    "testing.loc[0, 'CHR_ID'] = 11\n",
    "testing.loc[0, 'CHR_POS'] = 104942868"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = set(annotations.SYMBOL)\n",
    "more_genes = ['BCL-X', 'CTSB', 'NLRX1', 'ASC', 'POP1', 'IPAF', 'NLRC4', 'CARD18', 'NLRP3', 'NLRP1', 'HSP90AA1', 'MEFV', 'TNFSF14', 'NOD1', 'CASP1', 'SUGT1', 'NLRP5', 'TNF', 'TNFSF4', 'CD40LG', 'CARD4', 'PSTPIP1', 'NOD2', 'ICEBERG', 'AIM2', 'CIITA', 'NLRP12', 'TRA1', 'TNFSF11', 'BCL2L1', 'HSP90AB1', 'NLRP9', 'NLRP6', 'TNFSF5', 'NAIP', 'OX40L', 'NLRC5', 'BCL2', 'NLRP4', 'PYCARD', 'PYDC1', 'ICE', 'IPAF', 'HSP90B1', 'CASP5', 'BIRC1', 'NCRP1', 'AIM', 'NLR', 'IL1B', 'IL18', 'NLRC1', 'NLRC2', 'NLRC3', 'NLRC4', 'NLRC5', 'PYD', 'ASC', 'NAIP1', 'NAIP2', 'NAIP3', 'NAIP4', 'NAIP5', 'NAIP6', 'NACHT', 'PYCARD']\n",
    "genes.update(more_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlap(genes, df, window_radius=50000):\n",
    "    no_region_data = []\n",
    "    found = []\n",
    "    not_found = []\n",
    "    for gene in set(genes):\n",
    "        try:\n",
    "            info = get_region_info_for_gene(gene)\n",
    "        except:\n",
    "#             print(f'Error getting region info for {gene}')\n",
    "            no_region_data.append(gene)\n",
    "            continue\n",
    "        start = float(info['start'])\n",
    "        end = float(info['end'])\n",
    "        if info['chrom'] == 'X':\n",
    "            continue\n",
    "        chrom = float(info['chrom'])\n",
    "        i = chrom == df.CHR_ID\n",
    "        i &= (start - window_radius) <= df.CHR_POS\n",
    "        i &= (end + window_radius) >= df.CHR_POS\n",
    "#         print(gene, chrom, start, end)\n",
    "        n = sum(i)\n",
    "        if n:\n",
    "            found.append((gene, str(n)))\n",
    "        else:\n",
    "            not_found.append(gene)\n",
    "    print('No region data found:')\n",
    "    print(', '.join(no_region_data))\n",
    "    \n",
    "    print('\\nNo overlap:')\n",
    "    print(', '.join(not_found))\n",
    "    \n",
    "    print('\\nFound:')\n",
    "    print('\\n'.join([' '.join(x) for x in found]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No region data found:\n",
      "TRA1, NLRC1, ASC, IPAF, ICE, NAIP4, NCRP1, NACHT, POP1, NAIP1, NAIP6, PYD, OX40L, NLR, CARD4, BCL-X, AIM, NAIP5, NAIP3, NLRC2, NAIP2, BIRC1, ICEBERG\n",
      "\n",
      "No overlap:\n",
      "PYDC1, NAIP, HSP90AB1, BCL2, CIITA, NLRP3, DHX33, SUGT1, CASP5, AIM2, NLRP6, PSTPIP1, HSP90AA1, NLRP4, IL1B, NOD1, NLRX1, NLRP5, CTSB, NLRC4, TNFSF4, NLRP12, NLRC5, HSP90B1, PYCARD, CARD8, TNF, NLRP9, CARD18, TNFSF14, MEFV, GSDMD, TNFSF11, CASP1, NLRC3, IL18, CASP4, BCL2L1, NLRP1\n",
      "\n",
      "Found:\n",
      "[('NOD2', '1')]\n",
      "NOD2 1\n"
     ]
    }
   ],
   "source": [
    "get_overlap(genes, ibd)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
